backends/cuda-gen/ceed-cuda-gen-operator.c

```diff
@@ -32,6 +32,81 @@ static int CeedOperatorDestroy_Cuda_gen(CeedOperator op) {
   return CEED_ERROR_SUCCESS;
 }
 
+static int Waste(int threads_per_sm, int warp_size, int threads_per_elem,
+                 int elems_per_block) {
+  int useful_threads_per_block = threads_per_elem * elems_per_block;
+  // round up to nearest multiple of warp_size
+  int block_size = ((useful_threads_per_block + warp_size - 1) / warp_size) *
+                   warp_size;
+  int blocks_per_sm = threads_per_sm / block_size;
+  return threads_per_sm - useful_threads_per_block * blocks_per_sm;
+}
+
+// Choose the least wasteful block size constrained by blocks_per_sm of
+// max_threads_per_block.
+//
+// The x and y part of block[] contains per-element sizes (specified on input)
+// while the z part is number of elements.
+//
+// Problem setting: we'd like to make occupancy high with relatively few
+// inactive threads. CUDA (cuOccupancyMaxPotentialBlockSize) can tell us how
+// many threads can run.
+//
+// Note that full occupancy sometimes can't be achieved by one thread block. For
+// example, an SM might support 1536 threads in total, but only 1024 within a
+// single thread block. So cuOccupancyMaxPotentialBlockSize may suggest a block
+// size of 768 so that two blocks can run, versus one block of 1024 will prevent
+// a second block from running. The cuda-gen kernels are pretty heavy with lots
+// of instruction-level parallelism (ILP) so we'll generally be okay with
+// relatvely low occupancy and smaller thread blocks, but we solve a reasonably
+// general problem here. Empirically, we find that blocks bigger than about 256
+// have higher latency and worse load balancing when the number of elements is
+// modest.
+//
+// cuda-gen can't choose block sizes arbitrarily; they need to be a multiple of
+// the number of quadrature points (or number of basis functions). They also
+// have a lot of __syncthreads(), which is another point against excessively
+// large thread blocks. Suppose I have elements with 7x7x7 quadrature points.
+// This will loop over the last dimension, so we have 7*7=49 threads per
+// element. Suppose we have two elements = 2*49=98 useful threads. CUDA
+// schedules in units of full warps (32 threads), so 128 CUDA hardware threads
+// are effectively committed to that block. Now suppose
+// cuOccupancyMaxPotentialBlockSize returned 352. We can schedule 2 blocks of
+// size 98 (196 useful threads using 256 hardware threads), but not a third
+// block (which would need a total of 384 hardware threads).
+//
+// If instead, we had packed 3 elements, we'd have 2*49=147 useful threads
+// occupying 160 slots, and could schedule two blocks. Alterntively, we could
+// pack a single block of 7 elements (2*49=343 useful threads) into the 354
+// slots. The latter has the least "waste", but __syncthreads()
+// over-synchronizes and it might not pay off relative to smaller blocks.
```

Comment:
@YohannDudouit @nbeams Does this description of the performance tradeoffs for cuda-gen sound right to you? I did some profiling and viewing locally it it's how I would explain it.

Reply:
Makes sense to me.

Reply:
It's great, I think it describes things well.